tests/test_manifests.py

```diff
@@ -55,56 +71,32 @@ def test_product_manifest_properties():
-    # Eventually, we should also check the actual manifest content is valid SPDX data
-    # Then most of below can go away
-
-    # One component linked to this product
-    num_components = len(stream.get_latest_components())
-    assert num_components == 1
-
-    # Manifest contains info for all components + the product itself
-    assert len(manifest["packages"]) == 2
-
-    # Last "component" is actually the product
-    component_data = manifest["packages"][0]
-    product_data = manifest["packages"][-1]
-
-    # UUID and PURL, for each component attached to product, should be included in manifest
-    assert component_data["SPDXID"] == f"SPDXRef-{component.uuid}"
-    assert component_data["name"] == component.name
-    assert component_data["externalRefs"][0]["referenceLocator"] == component.purl
-
-    assert product_data["SPDXID"] == f"SPDXRef-{stream.uuid}"
-    assert product_data["name"] == stream.name
-    if stream.cpes:
-        assert product_data["externalRefs"][0]["referenceLocator"] == stream.cpes[0]
-    assert product_data["externalRefs"][-1]["referenceLocator"] == f"cpe:/{stream.ofuri}"
-
-    document_describes_product = {
-        "relatedSpdxElement": f"SPDXRef-{stream.uuid}",
-        "relationshipType": "DESCRIBES",
-        "spdxElementId": "SPDXRef-DOCUMENT",
-    }
-
-    component_is_package_of_product = {
-        "relatedSpdxElement": f"SPDXRef-{stream.uuid}",
-        "relationshipType": "PACKAGE_OF",
-        "spdxElementId": f"SPDXRef-{component.uuid}",
-    }
-
-    component_contains_nothing = {
-        "relatedSpdxElement": "NONE",
-        "relationshipType": "CONTAINS",
-        "spdxElementId": f"SPDXRef-{component.uuid}",
-    }
-
-    # For each component, one "component is package of product" relationship
-    # And one "component contains nothing" relationship
-    # Plus one "document describes product" relationship for the whole document at the end
-    assert len(manifest["relationships"]) == (num_components * 2) + 1
-    assert manifest["relationships"][0] == component_is_package_of_product
-    assert manifest["relationships"][1] == component_contains_nothing
-    assert manifest["relationships"][-1] == document_describes_product
+    component = Component.objects.get(pk=component.uuid)
+    assert len(component.provides) == 2
+    # print([name for name, _ in RelationshipType.__members__.items()])
+    print(stream.manifest)
+    err = StringIO()
+    parser = jsonparser.Parser(Builder(), FileLogger(err))
+    manifest_io = StringIO(stream.manifest)
+    doc, _ = parser.parse(manifest_io)
+
+    assert len(doc.packages) == 3
+
+    pkg_verif_code_node_error = "'None' is not a valid value for PKG_VERIF_CODE"
+    pkg_errors = []
+    for _ in range(len(doc.packages)):
+        pkg_errors.append(pkg_verif_code_node_error)
+    known_errors = "\n".join(pkg_errors)
+    errors = err.getvalue().strip()
+    assert errors == known_errors
```

Comment:
I think we probably shouldn't switch to this library until there are no more "known_errors" and TODOs. I wouldn't mind using a different library if there's one that works better. It would also be great if we could just do something like "assert doc.is_valid_spdx()" - not ure if the spdx-tools library has something like this.

Reply:
That's essentially what this library and this code is doing. However it needs some updating to work with SPDX 2.2 spec. I'll try and push some code to get it updated.